### PR TITLE
feat(interaction, module): separated lidar from veh updates, new Lidar module

### PR DESCRIPTION
**Pull request by FunKuchen**
## Description

The following changes were made on this branch

- A new `LidarUpdates `interaction was added
- The `ApplicationAmbassador ` was refactored to use this new interaction
- A new implementation of `LidarSensorModule `was created named `DefaultLidarSensorModule`
- The new `LidarSensorModule `is being used where LiDAR data was previously received through vehicle updates
- The `VehicleUnit ` now has the possibility to return the new LidarSensorModule, making LiDAR data accessible to units

- Not related to this, the gitignore now includes the ./tiles folder that saves downloaded osm sattelite/map data

## Issue(s) related to this PR

* Resolves internal issue 809
* Resolves internal issue 843
* Resolves internal issue 844
  
## Affected parts of the online documentation

* No changes required on website documentation
* Changes required on internal workshop slides, adressed in internal issue 952

## Definition of Done

**Prerequisites**

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] Your GitHub user id is linked with your Eclipse Account.

**Required**

- [x] The title of this merge request follows the scheme `type(scope): description`  (in the style of [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary))
- [ ] You have assigned a suitable label to this pull request (e.g., `enhancement`, or `bugfix`)
- [x] `origin/main` has been merged into your Fork.
- [x] Coding guidelines have been followed (see CONTRIBUTING.md).
- [ ] All checks on GitHub pass.
- [ ] All tests on [Jenkins](https://ci.eclipse.org/mosaic/job/mosaic/) pass.

**Requested** (can be enforced by maintainers)

- [ ] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [ ] If a bug has been fixed, a new unit test has been written (beforehand) to prove misbehavior
- [ ] There are no new SpotBugs warnings.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public void disable() {
+        this.enabled = false;
+    }
+
+    @Override
+    public void reactOnSensorUpdate(Consumer<PointCloud> callback) {
+        this.callback = callback;
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java:71]: Maybe allow more than one callback here (make this.callback a list)

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    @Override
+    public void disable() {
+        this.enabled = false;
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java:66]: Should we send a VehcielSensorActivation here to deactivate Lidar scanning in the simulators?

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        LidarUpdates other = (LidarUpdates) obj;
+        return new EqualsBuilder()
+                .append(this.nextUpdate, other.nextUpdate)
+                .append(this.updated, other.updated)
+                .isEquals();
+    }
+
+    //TODO what does this do?
```
Comment (kschrab) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:88]: This is for logging/printing the interaction.

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Time at which the next sensor update will be sent.
+     */
+    private long nextUpdate;
+
+    /**
+     * List of {@link PointCloud} containing LiDAR data from the simulator.
+     */
+    private Map<String, PointCloud> updated = new HashMap<>();
```
Comment (kschrab) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:46]: Why do we need a map here? Could be a list instead.

```
List<Pair<String, PointCloud>>
```

or

```
public record LidarUpdate(String vehicleOd, PointCloud pointCloud) {}

List<LidarUpdates.LidarUpdate> updates;
```

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                .append(this.updated, other.updated)
+                .isEquals();
+    }
+
+    //TODO what does this do?
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this, SHORT_PREFIX_STYLE)
+                .appendSuper(super.toString())
+                .append("updated", this.updated)
```
Comment (kschrab) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:93]: Don't just append the map here. I'm not even sure what happens if you just put the map here, I guess it would just print `HashMap@a78f8e9b` or something like that. Better append something meaningful, such as 

```java
  return new ToStringBuilder(this, SHORT_PREFIX_STYLE)
                .appendSuper(super.toString())
                .append("updated", "Updated Lidar data for vehicles [" + /*TODO collect vehicle ids*/ + "]);

**Review by kschrab — CHANGES_REQUESTED** [review]

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Override
+    public PointCloud getPointCloud() {
+        if (!enabled) {
+            return null;
+        }
+        return this.currentPointcloud;
+    }
+
+    public void addLidarUpdate(PointCloud pointCloud) {
```
Comment (kschrab) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java:82]: Maybe rename to `updatePointCloud`. `add..` sounds like that this module stores more than one pointcloud object.

File: .gitignore [hunk 7]
```diff
@@ -62,7 +62,7 @@ buildNumber.properties
 
 # MOSAIC
 
-tmp/
```
Comment (schwepmo) [inline .gitignore:65]: why is this removed?

Comment (schwepmo) [inline .gitignore:65]: why is this removed?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java [hunk 8]
```diff
@@ -693,6 +697,24 @@ private void process(final VehicleUpdates vehicleUpdates) {
         addEvent(triggerGarbageCollection);
     }
 
+    private void process(final LidarUpdates lidarUpdates) {
+        for (LidarUpdates.LidarUpdate lidarUpdate :lidarUpdates.getUpdated()) {
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java:701]: there is a space missing after the `:`

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public final static String TYPE_ID = createTypeIdentifier(LidarUpdates.class);
+
+    /**
+     * Time at which the next sensor update will be sent.
+     */
+    private long nextUpdate;
+
+
+    public record LidarUpdate(String unitId, PointCloud pointCloud) {}
```
Comment (schwepmo) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:44]: What do we think about this style-wise @kschrab? Typically, we placed in-line classes at the bottom of a class what should we do with records?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/VehicleUnit.java [hunk 10]
```diff
@@ -99,6 +103,7 @@ public VehicleUnit(String vehicleName, VehicleType vehicleType, final GeoPoint i
         }
 
         basicSensorModule = new EnvironmentBasicSensorModule();
+        defaultLidarSensorModule = new DefaultLidarSensorModule(this.getId());
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/VehicleUnit.java:106]: How would we switch to a non-default `LidarSensorModule`?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public void disable() {
+        this.enabled = false;
+        // Create a VehicleSensorActivation interaction to be sent to the RTI that disables the LiDAR sensor
+        VehicleSensorActivation interaction = new VehicleSensorActivation(
+                SimulationKernel.SimulationKernel.getCurrentSimulationTime(),
+                unitId,
+                0,
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java:73]: So deactivation just happens by setting the range to 0? Would a `VehicleSensorDeactivation` not cleaner?

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java (hunk 9 above)
Comment (kschrab) [reply lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:44]: >  Typically, we placed in-line classes at the bottom of a class

do we? 🤔

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/VehicleUnit.java (hunk 10 above)
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/VehicleUnit.java:106]: It's currently not intended to switch lidar modules, we just need an implementation of the API interface.

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java (hunk 9 above)
Comment (schwepmo) [reply lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:44]: At least that's what I did. I think at the top of the class makes a lot of sense too. Just not somewhere in between other declarations.

File: .gitignore (hunk 7 above)
Comment (FunKuchen) [reply .gitignore:65]: It was previously mentioned twice, see line 7

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java (hunk 11 above)
Comment (FunKuchen) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/DefaultLidarSensorModule.java:73]: Yes, I think so too. But as it is not implemented yet, maybe in a seperate issue?

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * String identifying the type of this interaction.
+     */
+    public final static String TYPE_ID = createTypeIdentifier(LidarUpdates.class);
+
+    /**
+     * Time at which the next sensor update will be sent.
+     */
+    private long nextUpdate;
```
Comment (kschrab) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:44]: Do we need this?

File: lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * List of {@link PointCloud} containing LiDAR data from the simulator.
+     */
+    List<LidarUpdate> updated;
```
Comment (kschrab) [inline lib/mosaic-interactions/src/main/java/org/eclipse/mosaic/interactions/environment/LidarUpdates.java:49]: make private

**Review by kschrab — CHANGES_REQUESTED** [review]

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/util/EventNicenessPriorityRegister.java [hunk 14]
```diff
@@ -24,6 +24,7 @@ public class EventNicenessPriorityRegister {
     public final static long VEHICLE_ADDED = -99_999_900;
     public final static long VEHICLE_UPDATED = -99_999_800;
     public final static long VEHICLE_REMOVED = -99_999_700;
+    public final static long LIDAR_UPDATED = -99_999_650;
```
[the remark this answers is not in the document]
Comment (FunKuchen) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/util/EventNicenessPriorityRegister.java:27]: It should, but I noticed, that I don't use this in my new implementation anymore. I will try to include it once more, although it seems to work the expected way right now

**Review by kschrab — APPROVED** [review]
